### PR TITLE
Add `lines` parser

### PR DESCRIPTION
**Pull request by mavam**
This PR adds a `lines` parser that chops up its input at newlines.

### CODE REVIEW

File: libtenzir/builtins/formats/lines.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        //        or last_finish + defaults::import::batch_timeout < now)
+        //
+        // To be discussed in review.
+        if (builder.rows() == 10) {
+          co_yield builder.finish();
+          builder = make_builder();
+        }
+      }
+      if (builder.rows() > 0)
+        co_yield builder.finish();
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/lines.cpp:107]:
```suggestion
        co_yield builder.finish("tenzir.lines");
```

For consistency with other operators and formats that have a schema without a name.

File: libtenzir/builtins/formats/lines.cpp [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+};
+
+class plugin final : public virtual parser_plugin<lines_parser> {
+public:
+  auto parse_parser(parser_interface& p) const
+    -> std::unique_ptr<plugin_parser> override {
+    auto parser = argument_parser{
+      name(), fmt::format("https://docs.tenzir.com/docs/formats/{}", name())};
+    auto args = parser_args{};
+    parser.add("-f,--field-name", args.field_name, "<string>");
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/lines.cpp:129]: I'm not sure this should be configurable at this level. Rename and put likely can already do this, so it's more or less duplicate functionality.

File: libtenzir/builtins/formats/lines.cpp (hunk 1 above)
Comment (dominiklohmann) [reply libtenzir/builtins/formats/lines.cpp:107]: Ah nvm, you're not using the adaptive builder and instead have a fixed schema above.

File: libtenzir/builtins/formats/lines.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            .hint("{}", *line)
+            .emit(ctrl.diagnostics());
+          co_return;
+        }
+        // FIXME: make configurable. Like in PCAP?
+        //
+        //    if (builder.rows() >= defaults::import::table_slice_size
+        //        or last_finish + defaults::import::batch_timeout < now)
+        //
+        // To be discussed in review.
```
[the remark this answers is not in the document]
Comment (dominiklohmann) [reply libtenzir/builtins/formats/lines.cpp:100]: We have a separate story for generalizing timeout and size limit handling with builders in operators. The simple answer is that we just don't know yet. But it's also not yet something we need to tackle.

File: libtenzir/builtins/formats/lines.cpp (hunk 2 above)
Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp:129]: No strong opinion here. What should be the default then? I can think of:

1. `line`
2. `data`

File: libtenzir/builtins/formats/lines.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp:100]: Should I copy the logic from the `pcap` parser for now?

File: libtenzir/builtins/formats/lines.cpp (hunk 2 above)
Comment (dominiklohmann) [reply libtenzir/builtins/formats/lines.cpp:129]: I like `line` as well.

File: libtenzir/builtins/formats/lines.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+            co_yield {};
+            continue;
+          }
+        } else {
+          ++num_non_empty_lines;
+        }
+        auto num_lines = skip_empty ? num_non_empty_lines
+                                    : num_non_empty_lines + num_empty_lines;
+        if (not builder.add(*line)) {
+          diagnostic::error("failed add line")
```
Comment (dominiklohmann) [inline libtenzir/builtins/formats/lines.cpp:89]:
```suggestion
          diagnostic::error("failed to add line")
```

**Review by dominiklohmann — APPROVED** [review]
Approving modulo the already open comments.

File: libtenzir/builtins/formats/lines.cpp [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+  template <class Inspector>
+  friend auto inspect(Inspector& f, parser_args& x) -> bool {
+    return f.object(x)
+      .pretty_name("parser_args")
+      .fields(f.field("field_name", x.field_name),
+              f.field("skip_empty", x.skip_empty));
+  }
+};
+
+auto line_type(std::string field_name = "data") -> type {
```
Comment (jachris) [inline libtenzir/builtins/formats/lines.cpp:37]:
```suggestion
auto line_type(std::string field_name) -> type {
```

**Review by jachris — COMMENTED** [review]
I can see that this can be quite useful to have!

File: libtenzir/builtins/formats/lines.cpp (hunk 4 above)
Comment (jachris) [reply libtenzir/builtins/formats/lines.cpp:89]: We could also just assert in such cases. I mean, this should never fail, and we don't have to do so much bookkeeping then.

Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp:89]: What's the maximum line size?

File: libtenzir/builtins/formats/lines.cpp (hunk 2 above)
Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp]: Removed the configurability of the field name in d4a7c856df and renamed the field to `line`.

File: libtenzir/builtins/formats/lines.cpp (hunk 4 above)
Comment (jachris) [reply libtenzir/builtins/formats/lines.cpp:80]: Good question. Right now, it's more or less unlimited. But that's probably fine? If the input has huge lines, this is what you want. And if it has no newlines at all, you are screwed anyway.

File: libtenzir/builtins/formats/lines.cpp [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    if (args.field_name and args.field_name->inner.empty())
+      diagnostic::error("field name must not be empty")
+        .primary(args.field_name->source)
+        .throw_();
+    return std::make_unique<lines_parser>(std::move(args));
+  }
+
+  auto name() const -> std::string override {
+    return "lines";
+  }
```
[the remark this answers is not in the document]
Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp]: Oh, right. Thanks! Removed in dcaa2c9de9.

File: libtenzir/builtins/formats/lines.cpp [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+};
+
+auto line_type(std::string field_name = "data") -> type {
+  TENZIR_ASSERT(not field_name.empty());
+  return type{
+    "tenzir.line",
+    record_type{
+      {std::move(field_name), string_type{}},
+    },
```
[the remark this answers is not in the document]
Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp]: With series, this should probably be a `series<string>`. Until then, the best we have is a record with a single string. So yeah, I think this means "I agree"?!

Feel free to resolve. 🙂

File: libtenzir/builtins/formats/lines.cpp (hunk 4 above)
Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp:80]: So should we then keep the error handling, if it *could* fail beyond 2GB lines?

I don't think it's realistic, but the error handling is there now.

Comment (jachris) [reply libtenzir/builtins/formats/lines.cpp:80]: Seems like Arrow just does a `static_cast<int32_t>(str.size())`, so... If we want to handle this, we must add some additional logic to `append_builder(...)`.

Comment (dominiklohmann) [reply libtenzir/builtins/formats/lines.cpp:80]: The limit Arrow has here is actually 2 GiB per array. This applies to all arrays. Individual arrays are stored in FlatBuffers tables, which have a 32 bit offset. FlatBuffers recently added a 64-bit builder, but that'd be a breaking change in Arrow's binary format.

Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp]: Welp, I'm still missing guidance on what to actually do here now.

File: libtenzir/builtins/formats/lines.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp]: Implemented in 3751e47c16.

File: libtenzir/builtins/formats/lines.cpp (hunk 4 above)
Comment (dominiklohmann) [reply libtenzir/builtins/formats/lines.cpp]: Keep the error handling as-is, and fix the grammar issue. You could add an `[[unlikely]]` if you want to prematurely optimize, but that doesn't matter much.

Comment (jachris) [reply libtenzir/builtins/formats/lines.cpp]: Agreed. Well, almost: You should remove (or truncate) the ` .hint("{}", *line)`.

Comment (mavam) [reply libtenzir/builtins/formats/lines.cpp]: Ah, right. I removed it.

**Pull request merged**